lib/Basic/FileSystem.cpp

```diff
@@ -62,6 +62,19 @@ class LocalFileSystem : public FileSystem {
     }
     return std::unique_ptr<llvm::MemoryBuffer>(result->release());
   }
+
+  virtual bool remove(const std::string& path) override {
+    struct stat statbuf;
+    if (sys::stat(path.c_str(), &statbuf) != 0) {
```

Comment:
You can still do the stat() after the failure if you want more fine grained switching after it fails... it's a bit more code but totally worth it for perf.